_overviews/core/programming-collections-generically.md

```diff
@@ -0,0 +1,320 @@
+
+object Gen {
+
+  /** Generator of `Int` values */
+  implicit def int: Gen[Int] =
+    new Gen[Int] { def get: Int = Random.nextInt() }
+
+  /** Generator of `Boolean` values */
+  implicit def boolean: Gen[Boolean] =
+    new Gen[Boolean] { def get: Boolean = Random.nextBoolean() }
+
+  /** Given a generator of `A` values, provides a generator of `List[A]` values */
+  implicit def list[A](implicit genA: Gen[A]): Gen[List[A]] =
+    new Gen[List[A]] {
+      def get: List[A] =
+        if (Random.nextInt(100) < 10) Nil
+        else genA.get :: get
+    }
+
+}
+~~~
+
+The last definition (`list`) generates a value of type `List[A]` given a generator
+of values of type `A`. We could implement a generator of `Vector[A]` or `Set[A]` as
+well, but their implementations would be very similar.
+
+Instead, we want to abstract over the type of the generated collection so that users
+can decide which collection type they want to produce.
+
+To achieve that we have to use `scala.collection.Factory`:
+
+~~~ scala
+trait Factory[-A, +C] {
+
+  /** @return A collection of type `C` containing the same elements
+    *         as the source collection `it`.
+    * @param it Source collection
+    */
+  def fromSpecific(it: IterableOnce[A]): C
+
+  /** Get a Builder for the collection. For non-strict collection
+    * types this will use an intermediate buffer.
+    * Building collections with `fromSpecific` is preferred
+    * because it can be lazy for lazy collections.
+    */
+  def newBuilder: Builder[A, C]
+}
+~~~
+
+The `Factory[A, C]` trait provides two ways of building a collection `C` from
+elements of type `A`:
+ 
+- `fromSpecific`, converts a source collection of `A` to a collection `C`,
+- `newBuilder`, provides a `Builder[A, C]`.
+
+The difference between these two methods is that the former does not necessarily
+evaluate the elements of the source collection. It can produce a non-strict
+collection type (such as `LazyList`) that does not evaluate its elements unless
+it is traversed. On the other hand, the builder-based way of constructing the
+collection necessarily evaluates the elements of the resulting collection.
+In practice, it is recommended to not eagerly evaluate the elements of the collection.
+
+Finally, here is how we can implement a generator of arbitrary collection types:
+
+~~~ scala
+import scala.collection.Factory
+
+implicit def collection[CC[_], A](implicit
+  genA: Gen[A],
+  factory: Factory[A, CC[A]]
+): Gen[CC[A]] =
+  new Gen[CC[A]] {
+    def get: CC[A] = {
+      val lazyElements =
+        LazyList.unfold(()) { _ =>
+          if (Random.nextInt(100) < 10) None
+          else Some((genA.get, ()))
+        }
+      factory.fromSpecific(lazyElements)
+    }
+  }
+~~~
+
+The implementation uses a lazy source collection of a random size (`lazyElements`).
+Then it calls the `fromSpecific` method of the `Factory` to build the collection
+expected by the user.
+
+Here is an example of use of `collection`:
```

Comment:
Not sure, but perhaps this whole section could begin with the example usage of `collection`, to show what we are trying to achieve, instead of saving it for an end reveal.

Reply:
meta: I seem to be making this suggestion repeatedly: go ahead, give the ending away right at the start, it's not a suspense movie :-)

each section should begin with a kind of abstract that gives away the entire section, then you follow with more detailed material, so the reader is free to decide whether to read the rest of the section or just skip to the next section. 80% of readers will be skipping through the text looking for something, relatively few readers will read the whole thing end to end like a novel.